### PR TITLE
lis2dw12: fix self test to match an5038 rev2

### LINKED ISSUE #1171: lis2dw12 self test
It utilizes fs 2g even though the device is in fs 4g
https://github.com/apache/mynewt-core/blob/26607457d6eb8db4918412834576784131f56794/hw/drivers/sensors/lis2dw12/src/lis2dw12.c#L1952

I also dont think it properly implements the app note. AN5038 Section 8 says the procedure consists of:
- enabling the accelerometer
- averaging five samples before enabling the self-test
- averaging five samples after enabling the self-test
- computing the difference in absolute value for each axis and verifying that it falls within
- a given range. The min and max values are provided in the datasheet.

The current one seems to add a positive reading to a negative one, and not do a ground truth reading at all.

### PR DESCRIPTION
**Pull request by jacobrosenthal**
This works with no changes from the previous patch on my freshly shipped hardware. Resubmitting. Closes #1171

### CODE REVIEW

File: hw/drivers/sensors/lis2dw12/src/lis2dw12.c [hunk 1]
```diff
@@ -1920,52 +1922,115 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
...
 
+    //discard
+    //TODO poll DRDY in STATUS (27h) instead?
+    rc = lis2dw12_get_data(itf, fs, &(data[0]), &(data[1]), &(data[2]));
+    if (rc) {
+        return rc;
+    }
+
+    /* take no st offset reading */
+    for(i=0; i<LIS2DW12_ST_NUM_READINGS; i++) {
```
Comment (mlaz) [inline hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: Why not `i = 0` instead of `i=0` ? You use both ways throughout the code.

Comment (mlaz) [reply hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: Although there is no specific rule for this one.

Comment (jacobrosenthal) [reply hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: Eh. I dont have a preference. Though the original file has i=0 in the lis2dw12_run_self_test function.

Thanks for the look, though Im mostly looking for a review from @vrahane as he previously had issues with this

**Comment by sterlinghughes** [comment]
Can we merge this?

**Comment by jacobrosenthal** [comment]
I would hate to merge on top of  @matthewwarnes  and @InternetofAwesome without some buy in from them

File: hw/drivers/sensors/lis2dw12/src/lis2dw12.c [hunk 2]
```diff
@@ -1920,52 +1922,115 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
...
+
+    /* wait 200ms */
+    os_time_delay(OS_TICKS_PER_SEC / 5 + 1);
+
+    rc = lis2dw12_get_fs(itf, &fs);
     if (rc) {
         return rc;
     }
 
+    //discard
```
Comment (kasjer) [inline hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: It is not clearly stated in application note, but test procedure will only work if FIFO is not enabled during self test described there.
To make this function work correctly I would suggest disabling FIFO during test, or make sure that FIFO does not have anything before reading self test data. Second method would take longer so disabling FIFO is batter IMHO.

File: hw/drivers/sensors/lis2dw12/src/lis2dw12.c [hunk 3]
```diff
@@ -1920,52 +1922,115 @@ int lis2dw12_get_int1_on_int2_map(struct sensor_itf *itf, uint8_t *val)
...
+        rc = lis2dw12_get_data(itf, fs, &(data[0]), &(data[1]), &(data[2]));
+        if (rc) {
+            return rc;
+        }
+        scratch[0] += data[0];
+        scratch[1] += data[1];
+        scratch[2] += data[2];
+    }
+
+    //average
```
Comment (kasjer) [inline hw/drivers/sensors/lis2dw12/src/lis2dw12.c]: C++ style comment introduced here, C-style was used in this file so why not stick to it.

**Comment by kasjer** [comment]
@jacobrosenthal I will provide separate PR to cover style and FIFO changes

**Pull request merged**